### PR TITLE
[#173432107], [#173436083] Add bonus activation lock

### PR DESCRIPTION
**Pull request by gunzip**
- removes the not working check on bonus orchestrator (which is scheduled by a message sent into a queue now)
- add a new model (BonusProcessing) to store processing bonuses for each user to lock the bonus activation procedure (only one bonus in status processing is permitted)
- the lockis relased when the start bonus activation orchestrator ends
- we use function bindings atm to read and write to cosmosdb

Needs a new collection (bonus-processing)

### CODE REVIEW

File: StartBonusActivation/handler.ts [hunk 1]
```diff
@@ -116,6 +118,14 @@ export function StartBonusActivationHandler(
           )
       )
       .fold(identity, apiBonusActivation => {
+        // Add the tuple (fiscalCode, bonusId) to the processing bonus collection
+        // this is used a s lock to avoid having more than one bonus in processing status
+        // The lock is removed when the orchestrator terminate
+        // tslint:disable-next-line: no-object-mutation
+        context.bindings.processingBonusIdOut = ProcessingBonus.encode({
```
Comment (gunzip) [inline StartBonusActivation/handler.ts:125]: here we write the lock (it's probably better to do this at the beginnig using the sdk; i'm gonna refactor this)

File: StartBonusActivationOrchestrator/handler.ts [hunk 2]
```diff
@@ -251,15 +253,24 @@ export const getStartBonusActivationOrchestratorHandler = (
...
         properties: {
-          id: operationId,
           name: "bonus.activation.error"
         }
       });
-      return false;
+    } finally {
+      // release user's lock when the orchestrator ends
+      yield context.df.callActivityWithRetry(
+        "ReleaseUserLockActivity",
```
Comment (gunzip) [inline StartBonusActivationOrchestrator/handler.ts:268]: here we remove the lock

File: EligibilityCheck/locks.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  IResponseErrorInternal,
+  ResponseErrorForbiddenNotAuthorized,
+  ResponseErrorInternal
+} from "italia-ts-commons/lib/responses";
+import { FiscalCode } from "italia-ts-commons/lib/strings";
+import { BonusCode } from "../generated/models/BonusCode";
+
+import * as t from "io-ts";
+
+export const ProcessingBonus = t.type({
```
Comment (cloudify) [inline EligibilityCheck/locks.ts:14]: `ProcessingBonusForUser` ?

Comment (cloudify) [reply EligibilityCheck/locks.ts:14]: there can be only one processing bonus for each user, right?

File: GetBonusActivation/handler.ts [hunk 4]
```diff
@@ -86,15 +75,15 @@ export function GetBonusActivationHandler(
...
-            "Not Found",
-            "Bonus activation not found"
+          // When the bonus is not found into the database but a new bonus activation
+          // is in progress for the user, we return 202 Accepted with the bonus id
+          return checkBonusActivationIsRunning(
+            context.bindings.processingBonusIdIn
+          ).fold<IGetBonusActivationHandlerOutput>(
+            response => response,
+            () =>
+              // Otherwise we return 4040 not found
```
Comment (cloudify) [inline GetBonusActivation/handler.ts:85]:
```suggestion
              // Otherwise we return 404 not found
```

File: GetBonusActivation/locks.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  ResponseErrorInternal,
+  ResponseSuccessAccepted
+} from "italia-ts-commons/lib/responses";
+import { FiscalCode, NonEmptyString } from "italia-ts-commons/lib/strings";
+import { InstanceId } from "../generated/definitions/InstanceId";
+import { BonusCode } from "../generated/models/BonusCode";
+
+import * as t from "io-ts";
+
+export const ProcessingBonus = t.type({
```
Comment (cloudify) [inline GetBonusActivation/locks.ts:15]: duplicate? the same type is declared in `EligibilityCheck/locks.ts`

File: EligibilityCheck/handler.ts [hunk 6]
```diff
@@ -48,23 +49,24 @@ export function EligibilityCheckHandler(): IEligibilityCheckHandler {
 
     // If a bonus activation for that user is in progress
     // returns 403 status response
-    const activationStatus = await client.getStatus(
-      makeStartBonusActivationOrchestratorId(fiscalCode)
+    const responseOrNotRunning = checkBonusActivationIsRunning(
```
Comment (francescopersico) [inline EligibilityCheck/handler.ts:52]:
```suggestion
    const responseOrBonusActivationNotRunning = checkBonusActivationIsRunning(
```

File: EligibilityCheck/locks.ts (hunk 3 above)
Comment (gunzip) [reply EligibilityCheck/locks.ts]: yes that's the reason of this PR (I've removed this type which is repeated anyway)

**Pull request merged**